faunadb-common/src/main/java/com/faunadb/common/models/tags/Tag.java

```diff
@@ -0,0 +1,84 @@
+package com.faunadb.common.models.tags;
+
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class Tag {
+
+    private static final int KEY_CHAR_LIMIT = 40;
+    private static final int VALUE_CHAR_LIMIT = 80;
+
+    private static final Pattern ALLOWABLE_CHARACTER_REGEX = Pattern.compile("^\\w+$");
+    private String key;
+    private String value;
+
+    public Tag() {
+        this.key = "";
+        this.value = "";
+    }
+
+    public Tag(String key, String value) {
+        this.key = getValidatedKey(key);
+        this.value = getValidatedValue(value);
+    }
+
+    @Override
+    public int hashCode() {
+        // For the purposes of hashing, two Tag objects with the same key should collide
+        return Objects.hash(this.key);
```

Comment:
this concerns me a little.

A client application using this class may want to produce a bunch of tags that have distinct values but the same key.

If we build equals and hashCode this way they would get unexpected behavior.

A couple questions.

1. Is there any reason user could not pass in multiple values for a key in a request?
2. If we want to ban that; I think we'll need another implementation unfortunately. Apologies, didn't think of this implication when I suggested we model tags.

Reply:
> A client application using this class may want to produce a bunch of tags that have distinct values but the same key

Keys should generally be unique - so I'd be interested in how a client expected to use them if they were not. You have an example?

1. Same as above, not sure how a customer would expect `Key1=Value1` and `Key1=Value2` to resolve...union? Overwriting (currently functionality) seems more straightforward to me.
2. Not at all, we would just make this `Objects.hash(this.key, this.value)` to make this unique across both key+value.

Reply:
> Not at all, we would just make this Objects.hash(this.key, this.value) to make this unique across both key+value.

I think we're getting our wires crossed.

I'm saying I think we should implement this differently than we have today ifyou want uniqueness by key.

> Same as above, not sure how a customer would expect Key1=Value1 and Key1=Value2 to resolve...union? Overwriting (currently functionality) seems more straightforward to me.

Imagine an app where i have a client and a collections of keys - but depending on which request i'm serving i want to give those keys different values.

Equals and Hashcode will not behave as i'd expect with this eimplementation.